source/common/upstream/outlier_detection_impl.cc

```diff
@@ -207,10 +232,54 @@ void DetectorImpl::onConsecutive5xxWorker(HostPtr host) {
   ejectHost(host, EjectionType::Consecutive5xx);
 }
 
+double Utility::srEjectionThreshold(double sr_sum, std::vector<double>& sr_data) {
+  double mean = sr_sum / sr_data.size();
+  double stdev = 0;
+  std::for_each(sr_data.begin(), sr_data.end(),
+                [&stdev, mean](double& v) { stdev += std::pow(v - mean, 2); });
+  stdev /= sr_data.size();
+  stdev = std::sqrt(stdev);
+
+  return mean - (SR_STDEV_FACTOR * stdev);
+}
+
 void DetectorImpl::onIntervalTimer() {
   SystemTime now = time_source_.currentSystemTime();
+  std::unordered_map<HostPtr, double> valid_sr_hosts;
+  std::vector<double> sr_data;
+  double sr_sum = 0;
   for (auto host : host_sinks_) {
     checkHostForUneject(host.first, host.second, now);
+
+    // Success Rate Outlier Detection.
+    // First swap out the current bucket been written to, to keep data valid.
+    host.second->updateCurrentSRBucket();
+
+    // If there are not enough hosts to begin with, don't do the work.
+    if (host_sinks_.size() >=
+        runtime_.snapshot().getInteger("outlier_detection.significant_host_threshold",
```

Comment:
Move this out. We don't need to call this everytime. Also, you can probably preallocate the double vector. Emplace back will get expensive for large asgs.